### PR TITLE
StringTerms.Bucket.getKeyAsNumber detection type

### LINKED ISSUE #28012: StringTerms.Bucket.getKeyAsNumber should detect integers
Currently it parses the string as a double, meaning it might lose accuracy if the value is an long which is greater than 2^52. We should try to detect whether the string represents a long or a double first.

### PR DESCRIPTION
**Pull request by matarrese**
This close #28012

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 1]
```diff
@@ -67,6 +67,10 @@ public Object getKey() {
         @Override
         public Number getKeyAsNumber() {
             // this method is needed for scripted numeric aggs
+            try {
+                return Long.parseLong(termBytes.utf8ToString());
+            } catch (NumberFormatException numberFormatException) {
+            }
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: I think a comment why this is being swallowed is in order here.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 2]
```diff
@@ -67,6 +67,10 @@ public Object getKey() {
         @Override
         public Number getKeyAsNumber() {
             // this method is needed for scripted numeric aggs
+            try {
+                return Long.parseLong(termBytes.utf8ToString());
+            } catch (NumberFormatException numberFormatException) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: I think it's best to name the exception as `ignored` here; it makes it clearer that the swallowing is intentional (and some IDEs complain about unused exceptions unless they are named `ignored`).

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a comment.

**Comment by jasontedor** [comment]
@matarrese In general it is best to add more context to your commit messages. A good commit message has a brief one-line summary (usually less than fifty characters), and then a more verbose message explaining the reasoning behind the change (wrapped to 72-column width). A more thorough explanation of this is [here](https://chris.beams.io/posts/git-commit/).

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 3]
```diff
@@ -67,6 +67,12 @@ public Object getKey() {
         @Override
         public Number getKeyAsNumber() {
             // this method is needed for scripted numeric aggs
+            try {
+                return Long.parseLong(termBytes.utf8ToString());
+            } catch (NumberFormatException ignored) {
+            }
+            //first the number is tried to be parsed as a Long
+            //if it fails, it will be parsed as a Double
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: This comment is a translation of the code to English, information that can already be gleaned by reading the code. The comment I am looking for here is *why* the code is written the way that it is written. *Why* are we doing this, why is it okay to swallow an exception?

**Review by jasontedor — CHANGES_REQUESTED** [review]
This comment is not sufficient.

**Comment by matarrese** [comment]
Hi @jasontedor thank you for all your comments. 
I was going to write some tests but I have just checked that this change (for both cases) is already well covered by the tests contained in these classes: `DoubleTermsIT.testScriptSingleValue` and `LongTermsIT.testScriptSingleValue`. 
Let me know what do you think about the update comment and the necessity to write some additional tests.

**Comment by jasontedor** [comment]
@matarrese This functionality can not be covered by existing tests; there is new code with different behavior than the what happened previously being added here?

**Comment by matarrese** [comment]
Please have a look at the newly added asserts

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 4]
```diff
@@ -67,6 +67,11 @@ public Object getKey() {
         @Override
         public Number getKeyAsNumber() {
             // this method is needed for scripted numeric aggs
+            try {
+                return Long.parseLong(termBytes.utf8ToString());
+            } catch (NumberFormatException ignored) {
+                //failed to be parsed as a Long, it will be parsed as a Double
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: This comment is still not explaining why we are doing this. It's again an English translation of the code. May I suggest going back to the original issue and reading why we are making this change? That is what we want to say here, the why.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left another comment about the comment.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java [hunk 5]
```diff
@@ -67,6 +67,11 @@ public Object getKey() {
         @Override
         public Number getKeyAsNumber() {
             // this method is needed for scripted numeric aggs
+            try {
+                return Long.parseLong(termBytes.utf8ToString());
+            } catch (NumberFormatException ignored) {
+                //failed to detect an Integer
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTerms.java]: @matarrese This is still not what we would want in a comment here. We want to explain why we doing this. Please look at the original issue that provoked this change and leave a comment here that explains the why behind first parsing as a long, and then leniently parsing as a double.

**Review by jasontedor — COMMENTED** [review]
I left another comment.

**Comment by matarrese** [comment]
@jasontedor please have a look at the new comment

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
Thanks @matarrese.

**Pull request merged**